services/gundeck/src/Gundeck/Redis.hs

```diff
@@ -86,42 +85,41 @@ connectRobust l retryStrategy connectLowLevel = do
           const $ Catch.Handler (\(e :: IOException) -> logEx (Log.err l) e "network error when connecting to Redis" >> pure True)
         ]
         . const -- ignore RetryStatus
-    reconnectRedis robustConnection = do
-      Log.info l $ Log.msg (Log.val "connecting to Redis")
-      conn <- connectLowLevel
-      Log.info l $ Log.msg (Log.val "successfully connected to Redis")
-
-      reconnectOnce <- once . retry $ reconnectRedis robustConnection -- avoid concurrent attempts to reconnect
```

Comment:
Essentially you say, `once` was not working? If `once` was working, there should never have been a race for updating the `MVar`. Or rather, the `MVar` would be swapped several times which could have been optimised by putting `swapMVar` below `once`. Or is there more to it which I haven't grasped in this case?

Reply:
What's bugging me is that I don't see the memory leak. After all pending Redis actions went through the (admittedly redundant) `swapMVar`, no memory should be retained. Though, I do see that it seems fixed by this PR.

Reply:
I also don't see it, which is why I spent 16 hours digging through heap profiles before I tried changing the code :blush: .  My best guess is that `once` isn't safe in a concurrent environment. It's implemented (indirectly) via `modifyMVar`, which comes with this warning:

> This function is only atomic if there are no other producers for this [MVar](https://hackage.haskell.org/package/base-4.18.0.0/docs/Control-Concurrent-MVar.html#t:MVar). In other words, it cannot guarantee that, by the time [modifyMVar_](https://hackage.haskell.org/package/base-4.18.0.0/docs/Control-Concurrent-MVar.html#v:modifyMVar_) gets the chance to write to the MVar, the value of the MVar has not been altered by a write operation from another thread.

which makes me think `once` is indeed the culprit. But I don't know!

The other difference between the two logics is that HEAD will still call `runRedis` in `runRobust` on a known-bad connection, and it could be some of the janky pipelining in hedis is responsible for leaking the stacks here.